File: apps/dashboard/src/features/enquiries/schemas.ts
import { z } from "zod";

import { enquiryPriorities, enquirySources, enquiryStatuses } from "./types";

const nullableDecimalSchema = z
  .union([z.string(), z.number()])
  .nullable()
  .transform((value) => (value === null ? null : String(value)));

export const enquiryNoteSchema = z.object({
  id: z.string(),
  author_id: z.string().nullable(),
  author_name: z.string().nullable(),
  note: z.string(),
  is_private: z.boolean(),
  created_at: z.string(),
});

const baseEnquiryFields = {
  id: z.string(),
  reference_code: z.string(),
  name: z.string(),
  email: z.string(),
  phone: z.string(),
  company_name: z.string(),
  source: z.enum(enquirySources),
  source_url: z.string(),
  status: z.enum(enquiryStatuses),
  priority: z.enum(enquiryPriorities),
  assigned_to_id: z.string().nullable(),
  assigned_to_name: z.string().nullable(),
  client_id: z.string().nullable(),
  lead_id: z.string().nullable(),
  submitted_at: z.string(),
  first_contacted_at: z.string().nullable(),
  resolved_at: z.string().nullable(),
  last_follow_up_at: z.string().nullable(),
  next_follow_up_at: z.string().nullable(),
  internal_summary: z.string(),
  loss_reason: z.string(),
  metadata: z.record(z.string(), z.unknown()),
  notes: z.array(enquiryNoteSchema),
  created_at: z.string(),
  updated_at: z.string(),
};

export const contactEnquirySchema = z.object({
  ...baseEnquiryFields,
  subject: z.string(),
  message: z.string(),
});

export const quoteEnquiryServiceSchema = z.object({
  id: z.string(),
  service_id: z.string(),
  service_title: z.string(),
  notes: z.string(),
  sort_order: z.number(),
});

export const quoteEnquirySchema = z.object({
  ...baseEnquiryFields,
  country: z.string(),
  website_url: z.string(),
  project_title: z.string(),
  project_description: z.string(),
  preferred_package_id: z.string().nullable(),
  budget_min: nullableDecimalSchema,
  budget_max: nullableDecimalSchema,
  budget_currency: z.string(),
  desired_start_date: z.string().nullable(),
  desired_completion_date: z.string().nullable(),
  quotation_id: z.string().nullable(),
  services: z.array(quoteEnquiryServiceSchema),
});

export const enquiryDashboardSchema = z.object({
  total_contact_enquiries: z.number(),
  total_quote_enquiries: z.number(),
  new_contact_enquiries: z.number(),
  new_quote_enquiries: z.number(),
  active_contact_enquiries: z.number(),
  active_quote_enquiries: z.number(),
  won_contact_enquiries: z.number(),
  won_quote_enquiries: z.number(),
  lost_contact_enquiries: z.number(),
  lost_quote_enquiries: z.number(),
  urgent_contact_enquiries: z.number(),
  urgent_quote_enquiries: z.number(),
  overdue_contact_follow_ups: z.number(),
  overdue_quote_follow_ups: z.number(),
  contact_enquiries_by_status: z.record(z.string(), z.number()),
  quote_enquiries_by_status: z.record(z.string(), z.number()),
  contact_enquiries_by_source: z.record(z.string(), z.number()),
  quote_enquiries_by_source: z.record(z.string(), z.number()),
});

File: backend/apps/enquiries/schemas.py
from datetime import date, datetime
from decimal import Decimal
from typing import Literal
from uuid import UUID

from ninja import Schema
from pydantic import Field


EnquiryStatusValue = Literal[
    "new",
    "assigned",
    "contacted",
    "qualified",
    "proposal_sent",
    "won",
    "lost",
    "spam",
    "archived",
]

EnquiryPriorityValue = Literal[
    "low",
    "normal",
    "high",
    "urgent",
]

EnquirySourceValue = Literal[
    "website",
    "google",
    "facebook",
    "instagram",
    "linkedin",
    "tiktok",
    "whatsapp",
    "referral",
    "email",
    "phone",
    "manual",
    "other",
]


PublicQuoteContactMethodValue = Literal[
    "email",
    "whatsapp",
]

PublicQuoteContactTimeValue = Literal[
    "",
    "morning",
    "afternoon",
    "evening",
    "anytime",
]

PublicQuoteSurfaceValue = Literal[
    "page",
    "modal",
]


class PublicQuoteRequestSchema(Schema):
    full_name: str
    company_name: str = ""
    service_required: str
    email: str
    whatsapp_number: str = ""
    preferred_contact_method: PublicQuoteContactMethodValue
    country: str
    project_description: str
    best_time_to_contact: PublicQuoteContactTimeValue = ""
    source_surface: PublicQuoteSurfaceValue = "page"
    source_url: str = ""


class PublicQuoteResponseSchema(Schema):
    status: str
    reference_code: str
    message: str


class ContactEnquiryCreateSchema(Schema):
    reference_code: str
    name: str
    email: str = ""
    phone: str = ""
    company_name: str = ""
    subject: str = ""
    message: str
    source: EnquirySourceValue = "website"
    source_url: str = ""
    priority: EnquiryPriorityValue = "normal"
    assigned_to_id: UUID | None = None
    client_id: UUID | None = None
    lead_id: UUID | None = None
    metadata: dict = Field(default_factory=dict)


class QuoteEnquiryServiceInputSchema(Schema):
    service_id: UUID
    notes: str = ""
    sort_order: int = 0


class QuoteEnquiryCreateSchema(Schema):
    reference_code: str
    name: str
    email: str = ""
    phone: str = ""
    company_name: str = ""
    country: str = ""
    website_url: str = ""
    project_title: str = ""
    project_description: str
    preferred_package_id: UUID | None = None
    budget_min: Decimal | None = None
    budget_max: Decimal | None = None
    budget_currency: str = "LKR"
    desired_start_date: date | None = None
    desired_completion_date: date | None = None
    source: EnquirySourceValue = "website"
    source_url: str = ""
    priority: EnquiryPriorityValue = "normal"
    assigned_to_id: UUID | None = None
    client_id: UUID | None = None
    lead_id: UUID | None = None
    metadata: dict = Field(default_factory=dict)
    services: list[
        QuoteEnquiryServiceInputSchema
    ] = Field(default_factory=list)


class EnquiryStatusUpdateSchema(Schema):
    status: EnquiryStatusValue
    loss_reason: str = ""


class EnquiryAssignmentSchema(Schema):
    assigned_to_id: UUID | None = None
    priority: EnquiryPriorityValue = "normal"
    internal_summary: str = ""
    next_follow_up_at: datetime | None = None


class EnquiryNoteCreateSchema(Schema):
    note: str
    is_private: bool = True


class EnquiryNoteSchema(Schema):
    id: UUID
    author_id: UUID | None
    author_name: str | None
    note: str
    is_private: bool
    created_at: datetime


class QuoteEnquiryServiceSchema(Schema):
    id: UUID
    service_id: UUID
    service_title: str
    notes: str
    sort_order: int


class ContactEnquirySchema(Schema):
    id: UUID
    reference_code: str
    name: str
    email: str
    phone: str
    company_name: str
    subject: str
    message: str
    source: str
    source_url: str
    status: str
    priority: str
    assigned_to_id: UUID | None
    assigned_to_name: str | None
    client_id: UUID | None
    lead_id: UUID | None
    submitted_at: datetime
    first_contacted_at: datetime | None
    resolved_at: datetime | None
    last_follow_up_at: datetime | None
    next_follow_up_at: datetime | None
    internal_summary: str
    loss_reason: str
    metadata: dict
    notes: list[EnquiryNoteSchema]
    created_at: datetime
    updated_at: datetime


class QuoteEnquirySchema(Schema):
    id: UUID
    reference_code: str
    name: str
    email: str
    phone: str
    company_name: str
    country: str
    website_url: str
    project_title: str
    project_description: str
    preferred_package_id: UUID | None
    budget_min: Decimal | None
    budget_max: Decimal | None
    budget_currency: str
    desired_start_date: date | None
    desired_completion_date: date | None
    source: str
    source_url: str
    status: str
    priority: str
    assigned_to_id: UUID | None
    assigned_to_name: str | None
    client_id: UUID | None
    lead_id: UUID | None
    quotation_id: UUID | None
    submitted_at: datetime
    first_contacted_at: datetime | None
    resolved_at: datetime | None
    last_follow_up_at: datetime | None
    next_follow_up_at: datetime | None
    internal_summary: str
    loss_reason: str
    metadata: dict
    services: list[QuoteEnquiryServiceSchema]
    notes: list[EnquiryNoteSchema]
    created_at: datetime
    updated_at: datetime



class ContactEnquiryUpdateSchema(
    ContactEnquiryCreateSchema
):
    pass


class QuoteEnquiryUpdateSchema(
    QuoteEnquiryCreateSchema
):
    pass


class EnquiryFollowUpSchema(Schema):
    next_follow_up_at: datetime | None = None


class EnquiryDashboardSchema(Schema):
    total_contact_enquiries: int
    total_quote_enquiries: int
    new_contact_enquiries: int
    new_quote_enquiries: int
    active_contact_enquiries: int
    active_quote_enquiries: int
    won_contact_enquiries: int
    won_quote_enquiries: int
    lost_contact_enquiries: int
    lost_quote_enquiries: int
    urgent_contact_enquiries: int
    urgent_quote_enquiries: int
    overdue_contact_follow_ups: int
    overdue_quote_follow_ups: int
    contact_enquiries_by_status: dict
    quote_enquiries_by_status: dict
    contact_enquiries_by_source: dict
    quote_enquiries_by_source: dict
